From 1d5f6069f6212c51aa56b64922e60ec4eb3c0e3a Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Tue, 24 Dec 2024 16:47:53 +0800
Subject: [PATCH 17/23] Revert "cgroup: Make operations on the cgroup root_list
 RCU safe"

hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/IBDFAV

----------------------------------------------------------------------

This reverts commit 486eff6ae48b78edc20a4e51596f332c20f2f75f.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 include/linux/cgroup-defs.h     |  1 -
 kernel/cgroup/cgroup-internal.h |  3 +--
 kernel/cgroup/cgroup.c          | 23 +++++++----------------
 3 files changed, 8 insertions(+), 19 deletions(-)

diff --git a/include/linux/cgroup-defs.h b/include/linux/cgroup-defs.h
index 05ece896af7d..6e3227a688de 100644
--- a/include/linux/cgroup-defs.h
+++ b/include/linux/cgroup-defs.h
@@ -591,7 +591,6 @@ struct cgroup_root {
 
 	/* A list running through the active hierarchies */
 	struct list_head root_list;
-	struct rcu_head rcu;
 
 	/* Hierarchy-specific flags */
 	unsigned int flags;
diff --git a/kernel/cgroup/cgroup-internal.h b/kernel/cgroup/cgroup-internal.h
index f5fb12890645..96a9bd2c26f0 100644
--- a/kernel/cgroup/cgroup-internal.h
+++ b/kernel/cgroup/cgroup-internal.h
@@ -170,8 +170,7 @@ extern struct list_head cgroup_roots;
 
 /* iterate across the hierarchies */
 #define for_each_root(root)						\
-	list_for_each_entry_rcu((root), &cgroup_roots, root_list,	\
-				lockdep_is_held(&cgroup_mutex))
+	list_for_each_entry((root), &cgroup_roots, root_list)
 
 /**
  * for_each_subsys - iterate all enabled cgroup subsystems
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index c26a9b3a3576..52fe6ba2fefd 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -1315,7 +1315,7 @@ static void cgroup_exit_root_id(struct cgroup_root *root)
 
 void cgroup_free_root(struct cgroup_root *root)
 {
-	kfree_rcu(root, rcu);
+	kfree(root);
 }
 
 static void cgroup_destroy_root(struct cgroup_root *root)
@@ -1348,7 +1348,7 @@ static void cgroup_destroy_root(struct cgroup_root *root)
 	spin_unlock_irq(&css_set_lock);
 
 	if (!list_empty(&root->root_list)) {
-		list_del_rcu(&root->root_list);
+		list_del(&root->root_list);
 		cgroup_root_count--;
 	}
 
@@ -1388,15 +1388,7 @@ static inline struct cgroup *__cset_cgroup_from_root(struct css_set *cset,
 		}
 	}
 
-	/*
-	 * If cgroup_mutex is not held, the cgrp_cset_link will be freed
-	 * before we remove the cgroup root from the root_list. Consequently,
-	 * when accessing a cgroup root, the cset_link may have already been
-	 * freed, resulting in a NULL res_cgroup. However, by holding the
-	 * cgroup_mutex, we ensure that res_cgroup can't be NULL.
-	 * If we don't hold cgroup_mutex in the caller, we must do the NULL
-	 * check.
-	 */
+	BUG_ON(!res_cgroup);
 	return res_cgroup;
 }
 
@@ -1455,6 +1447,7 @@ static struct cgroup *current_cgns_cgroup_dfl(void)
 static struct cgroup *cset_cgroup_from_root(struct css_set *cset,
 					    struct cgroup_root *root)
 {
+	lockdep_assert_held(&cgroup_mutex);
 	lockdep_assert_held(&css_set_lock);
 
 	return __cset_cgroup_from_root(cset, root);
@@ -1462,9 +1455,7 @@ static struct cgroup *cset_cgroup_from_root(struct css_set *cset,
 
 /*
  * Return the cgroup for "task" from the given hierarchy. Must be
- * called with css_set_lock held to prevent task's groups from being modified.
- * Must be called with either cgroup_mutex or rcu read lock to prevent the
- * cgroup root from being destroyed.
+ * called with cgroup_mutex and css_set_lock held.
  */
 struct cgroup *task_cgroup_from_root(struct task_struct *task,
 				     struct cgroup_root *root)
@@ -2039,7 +2030,7 @@ void init_cgroup_root(struct cgroup_fs_context *ctx)
 	struct cgroup_root *root = ctx->root;
 	struct cgroup *cgrp = &root->cgrp;
 
-	INIT_LIST_HEAD_RCU(&root->root_list);
+	INIT_LIST_HEAD(&root->root_list);
 	atomic_set(&root->nr_cgrps, 1);
 	cgrp->root = root;
 	init_cgroup_housekeeping(cgrp);
@@ -2123,7 +2114,7 @@ int cgroup_setup_root(struct cgroup_root *root, u16 ss_mask)
 	 * care of subsystems' refcounts, which are explicitly dropped in
 	 * the failure exit path.
 	 */
-	list_add_rcu(&root->root_list, &cgroup_roots);
+	list_add(&root->root_list, &cgroup_roots);
 	cgroup_root_count++;
 
 	/*
-- 
2.25.1

